File: Java/opentraceability/src/main/java/opentraceability/mappers/OTMappingTypeInformation.java
package opentraceability.mappers;

import opentraceability.utility.ListExtensions;
import opentraceability.utility.ReflectionUtility;
import opentraceability.utility.attributes.*;
import tangible.StringHelper;

import java.lang.reflect.Field;

import java.util.*;
import java.util.concurrent.locks.ReentrantLock;

public class OTMappingTypeInformation {
    public Class Class;
    public ArrayList<OTMappingTypeInformationProperty> properties = new ArrayList<>();
    public Map<String, OTMappingTypeInformationProperty> dic = new HashMap<>();
    public Field extensionKDEs = null;
    public Field extensionAttributes = null;

    public OTMappingTypeInformation() { }

    public OTMappingTypeInformation(Class<?> type, EPCISDataFormat format, Boolean isMasterDataMapping) {
        for (Field prop : type.getFields()) {
            if (prop != null) {
                if (format == EPCISDataFormat.XML && ReflectionUtility.getFieldAnnotation(prop, OpenTraceabilityXmlIgnoreAttribute.class) != null) {
                    continue;
                }

                if (isMasterDataMapping) {
                    List<OpenTraceabilityMasterDataAttribute> mdAtt = ReflectionUtility.getFieldAnnotations(prop, OpenTraceabilityMasterDataAttribute.class);
                    if (mdAtt != null && mdAtt.size() > 0) {
                        OTMappingTypeInformationProperty property = new OTMappingTypeInformationProperty(prop, mdAtt.get(0), format);
                        this.properties.add(property);
                        dic.put(property.name, property);
                    }
                }
                else
                {
                    List<OpenTraceabilityAttribute> atts = ReflectionUtility.getFieldAnnotations(prop, OpenTraceabilityAttribute.class);
                    List<OpenTraceabilityJsonAttribute> jsonAtt = ReflectionUtility.getFieldAnnotations(prop, OpenTraceabilityJsonAttribute.class);
                    List<OpenTraceabilityProductsAttribute> productAtts = ReflectionUtility.getFieldAnnotations(prop, OpenTraceabilityProductsAttribute.class);

                    if (atts.size() > 0) {
                        for (OpenTraceabilityAttribute att : atts) {
                            OTMappingTypeInformationProperty property = new OTMappingTypeInformationProperty(prop, att, format);
                            if (!dic.containsKey(property.name)) {
                                this.properties.add(property);
                                dic.put(property.name, property);
                                if (!StringHelper.isNullOrEmpty(att.ns()))
                                {
                                    String key = "{" + att.ns() + "}" + att.name();
                                    dic.put(key, property);
                                }
                            }
                        }
                    } else if (jsonAtt != null && jsonAtt.size() > 0 && format == EPCISDataFormat.JSON) {
                        OTMappingTypeInformationProperty property = new OTMappingTypeInformationProperty(prop, jsonAtt.get(0), format);
                        if (!dic.containsKey(property.name)) {
                            this.properties.add(property);
                            dic.put(property.name, property);
                        }
                    } else if (productAtts.size() > 0) {
                        for (var att : productAtts) {
                            OTMappingTypeInformationProperty property = new OTMappingTypeInformationProperty(prop, att, format);
                            if (!dic.containsKey(property.name)) {
                                this.properties.add(property);
                                dic.put(property.name, property);
                            }
                        }
                    }
                    else if (ReflectionUtility.getFieldAnnotation(prop, OpenTraceabilityExtensionElementsAttribute.class) != null)
                    {
                        extensionKDEs = prop;
                    }
                    else if (ReflectionUtility.getFieldAnnotation(prop, OpenTraceabilityExtensionAttributesAttribute.class) != null)
                    {
                        extensionAttributes = prop;
                    }

//                    properties.sort((a, b) -> Boolean.compare(a.sequenceOrder == null, b.sequenceOrder == null) != 0
//                            ? a.sequenceOrder == null ? 1 : -1
//                            : a.sequenceOrder - b.sequenceOrder);
                }
            }
        }
    }

    private static final ReentrantLock locker = new ReentrantLock();
    private static final Map<Class, OTMappingTypeInformation> xmlTypeInfos = new HashMap<>();
    private static final Map<Class, OTMappingTypeInformation> jsonTypeInfos = new HashMap<>();
    private static final Map<Class, OTMappingTypeInformation> masterDataXmlTypeInfos = new HashMap<>();
    private static final Map<Class, OTMappingTypeInformation> masterDataJsonTypeInfos = new HashMap<>();

    public static OTMappingTypeInformation getXmlTypeInfo(Class<?> t) {
        if (!xmlTypeInfos.containsKey(t)) {
            locker.lock();
            if (!xmlTypeInfos.containsKey(t)) {
                OTMappingTypeInformation typeInfo = new OTMappingTypeInformation(t, EPCISDataFormat.XML, false);
                xmlTypeInfos.put(t, typeInfo);
            }
            locker.unlock();
        }
        return xmlTypeInfos.get(t);
    }

    public static OTMappingTypeInformation getJsonTypeInfo(Class t) {
        if (!jsonTypeInfos.containsKey(t)) {
            locker.lock();
            if (!jsonTypeInfos.containsKey(t)) {
                OTMappingTypeInformation typeInfo = new OTMappingTypeInformation(t, EPCISDataFormat.JSON, false);
                jsonTypeInfos.put(t, typeInfo);
            }
            locker.unlock();
        }
        return jsonTypeInfos.get(t);
    }

    public static OTMappingTypeInformation getMasterDataXmlTypeInfo(Class t) {
        if (!masterDataXmlTypeInfos.containsKey(t)) {
            locker.lock();
            if (!masterDataXmlTypeInfos.containsKey(t)) {
                OTMappingTypeInformation typeInfo = new OTMappingTypeInformation(t, EPCISDataFormat.XML, true);
                masterDataXmlTypeInfos.put(t, typeInfo);
            }
            locker.unlock();
        }
        return masterDataXmlTypeInfos.get(t);
    }

    public static OTMappingTypeInformation getMasterDataJsonTypeInfo(Class t) {
        if (!masterDataJsonTypeInfos.containsKey(t)) {
            locker.lock();
            if (!masterDataJsonTypeInfos.containsKey(t)) {
                OTMappingTypeInformation typeInfo = new OTMappingTypeInformation(t, EPCISDataFormat.JSON, true);
                masterDataJsonTypeInfos.put(t, typeInfo);
            }
            locker.unlock();
        }
        return masterDataJsonTypeInfos.get(t);
    }

    public OTMappingTypeInformationProperty get(String name, String ns)
    {
        var prop = dic.get(name);
        if (prop == null)
        {
            String localName = ListExtensions.LastOrDefault(Arrays.stream(name.split(":")));
            String key = "{" + ns + "}" + localName;
            prop = dic.get(key);
        }
        return prop;
    }
}

File: Java/OpenTraceability/src/main/opentraceability/mappers/otmappingtypeinformation.kt
package mappers

import utility.attributes.*
import java.beans.BeanInfo
import java.util.*
import java.lang.reflect.Type
import kotlin.reflect.KProperty

class OTMappingTypeInformation {
    companion object {
        var _XmlTypeInfos: MutableMap<Type, OTMappingTypeInformation> = mutableMapOf()
        var _JsonTypeInfos: MutableMap<Type, OTMappingTypeInformation> = mutableMapOf()
        var _masterDataXmlTypeInfos: MutableMap<Type, OTMappingTypeInformation> = mutableMapOf()
        var _masterDataJsonTypeInfos: MutableMap<Type, OTMappingTypeInformation> = mutableMapOf()

        fun GetXmlTypeInfo(t: KClass<*>): OTMappingTypeInformation {
            if (!_XmlTypeInfos.containsKey(t)) {
                synchronized(_locker) {
                    if (!_XmlTypeInfos.containsKey(t)) {
                        val typeInfo = OTMappingTypeInformation(t, EPCISDataFormat.XML)
                        _XmlTypeInfos[t] = typeInfo
                    }
                }
            }
            return _XmlTypeInfos[t]!!
        }

        fun GetJsonTypeInfo(t: KClass<*>): OTMappingTypeInformation {
            if (!_JsonTypeInfos.containsKey(t)) {
                synchronized(_locker) {
                    if (!_JsonTypeInfos.containsKey(t)) {
                        val typeInfo = OTMappingTypeInformation(t, EPCISDataFormat.JSON)
                        _JsonTypeInfos[t] = typeInfo
                    }
                }
            }
            return _JsonTypeInfos[t]!!
        }

        fun GetMasterDataXmlTypeInfo(t: KClass<*>): OTMappingTypeInformation {
            if (!_masterDataXmlTypeInfos.containsKey(t)) {
                synchronized(_locker) {
                    if (!_masterDataXmlTypeInfos.containsKey(t)) {
                        val typeInfo = OTMappingTypeInformation(t, EPCISDataFormat.XML, true)
                        _masterDataXmlTypeInfos[t] = typeInfo
                    }
                }
            }
            return _masterDataXmlTypeInfos[t]!!
        }

        fun GetMasterDataJsonTypeInfo(t: KClass<*>): OTMappingTypeInformation {
            if (!_masterDataJsonTypeInfos.containsKey(t)) {
                synchronized(_locker) {
                    if (!_masterDataJsonTypeInfos.containsKey(t)) {
                        val typeInfo = OTMappingTypeInformation(t, EPCISDataFormat.JSON, true)
                        _masterDataJsonTypeInfos[t] = typeInfo
                    }
                }
            }
            return _masterDataJsonTypeInfos[t]!!
        }

    }

    lateinit var Type: Type
    var Properties: ArrayList<OTMappingTypeInformationProperty> = ArrayList<OTMappingTypeInformationProperty>()
    var ExtensionKDEs: BeanInfo? = null
    var ExtensionAttributes: BeanInfo? = null

    class OTMappingTypeInformation(val type: KClass<*>, format: EPCISDataFormat, isMasterDataMapping: Boolean = false) {
        val properties: MutableList<OTMappingTypeInformationProperty> = mutableListOf()
        private val _dic: MutableMap<String, OTMappingTypeInformationProperty> = mutableMapOf()
        var extensionKDEs: KProperty<*>? = null
        var extensionAttributes: KProperty<*>? = null

        init {
            type.memberProperties.forEach { p ->
                if (format == EPCISDataFormat.XML && p.findAnnotation<OpenTraceabilityXmlIgnoreAttribute>() != null) {
                    return@forEach
                }

                if (isMasterDataMapping) {
                    val mdAtt = p.findAnnotation<OpenTraceabilityMasterDataAttribute>()

                    if (mdAtt != null) {
                        val property = OTMappingTypeInformationProperty(p, mdAtt, format)
                        properties.add(property)
                        _dic[property.Name] = property
                    }
                } else {
                    val atts = p.annotations.filterIsInstance<OpenTraceabilityAttribute>()
                    val jsonAtt = p.findAnnotation<OpenTraceabilityJsonAttribute>()
                    val productAtts = p.annotations.filterIsInstance<OpenTraceabilityProductsAttribute>()

                    if (atts.isNotEmpty()) {
                        atts.forEach { att ->
                            val property = OTMappingTypeInformationProperty(p, att, format)
                            if (!_dic.containsKey(property.Name)) {
                                properties.add(property)
                                _dic[property.Name] = property
                            }
                        }
                    } else if (jsonAtt != null && format == EPCISDataFormat.JSON) {
                        val property = OTMappingTypeInformationProperty(p, jsonAtt, format)
                        if (!_dic.containsKey(property.Name)) {
                            properties.add(property)
                            _dic[property.Name] = property
                        }
                    } else if (productAtts.isNotEmpty()) {
                        productAtts.forEach { att ->
                            val property = OTMappingTypeInformationProperty(p, att, format)
                            if (!_dic.containsKey(property.Name)) {
                                properties.add(property)
                                _dic[property.Name] = property
                            }
                        }
                    } else if (p.findAnnotation<OpenTraceabilityExtensionElementsAttribute>() != null) {
                        extensionKDEs = p
                    } else if (p.findAnnotation<OpenTraceabilityExtensionAttributesAttribute>() != null) {
                        extensionAttributes = p
                    }

                    properties.sortBy { p -> p.SequenceOrder == null }
                    properties.sortBy { p -> p.SequenceOrder }
                }
            }
        }

        operator fun get(name: String): OTMappingTypeInformationProperty? {
            return _dic[name]
        }
    }


}
